### PR TITLE
Make sure to use a dedicated metadata connection

### PR DESCRIPTION
**Pull request by onderkalaci**
With this commit, we make sure to use a dedicated connection per
node for all the metadata operations within the same transaction.

This is needed because the same metadata (e.g., metadata includes
the distributed table on the workers) can be modified accross
multiple connections.

With this connection we guarantee that there is a single metadata connection.
But note that this connection can be used for any other operation.
In other words, this connection is not only reserved for metadata
operations.

### CODE REVIEW

File: src/backend/distributed/connection/connection_management.c [hunk 1]
```diff
@@ -425,7 +439,6 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 		/* don't return claimed connections */
 		if (connection->claimedExclusively)
 		{
-			/* connection is in use for an ongoing operation */
```
Comment (onderkalaci) [inline src/backend/distributed/connection/connection_management.c:428]: todo: add the comment back

File: src/backend/distributed/connection/connection_management.c [hunk 2]
```diff
@@ -454,13 +467,97 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
...
+		 */
+		MultiConnection *metadataConnection =
+			dlist_container(MultiConnection, connectionNode,
+							dlist_head_node(connections));
+
+
+		/* remember that we use this connection for metadata operations */
+		metadataConnection->useForMetadataOperations = true;
+
+		#ifdef USE_ASSERT_CHECKING
```
Comment (marcocitus) [inline src/backend/distributed/connection/connection_management.c:518]: maybe just ifdef out the contents of the function

File: src/backend/distributed/connection/connection_management.c [hunk 3]
```diff
@@ -454,13 +467,97 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
...
 		}
 
+		if ((flags & REQUIRE_METADATA_CONNECTION) &&
+			!connection->useForMetadataOperations)
+		{
+			/*
+			 * The caller requested a metadata connection, and this is not the
+			 * metadata connection.
+			 */
+			continue;
```
Comment (marcocitus) [inline src/backend/distributed/connection/connection_management.c:477]: what if we don't have a metadata connection yet? can we use an existing one?

Comment (onderkalaci) [reply src/backend/distributed/connection/connection_management.c:477]: we handle that below, after the loop? We pick the first connection as the metadata connection, and mark `useForMetadataOperations`

File: src/backend/distributed/connection/connection_management.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/connection/connection_management.c:518]: I use this function in two different places, and the code is not very short. So, I have an inclination to keep like this if that makes sense

Comment (marcocitus) [reply src/backend/distributed/connection/connection_management.c:518]: Also maybe use Assry in the name.

**Review by marcocitus — APPROVED** [review]
Seems nice and simple. It's perhaps not ideal to get connections in different modes, but I think it's useful to track this.

**Comment by onderkalaci** [comment]
In theory, we should be backporting this change to 10.2. But given it is very unlikely to hit this bug in practice (as discussed in the PR), I'm NOT adding the backport label.

If needed, we can consider backporting.
